### PR TITLE
Add CIAM authority support

### PR DESCRIPTION
**Pull request by sameerag**
**Add CIAM support in MSAL JS**

Azure Active Directory (Azure AD) is adding support to customer identity access management (CIAM) solution that lets you create secure, customized sign-in experiences for your customer-facing apps and services. With these built-in CIAM features, Azure AD can serve as the identity provider and access management service for your customer scenarios.

> The CIAM authority format is still in the process of standardization. It is determined to be split from AAD and currently supported in the below formats:

* `https://tenantName.ciamlogin.com/tenant` 
* `https://tenantName.ciamlogin.com`

> All MSALs have currently agreed to add two rules for CIAM Authority:

1. Add a new `CIAM` authority type
2. If the i/p CIAM  authority does not have a path, append `"tenant.onmicrosoft.com"` to the PATH as a temporary measure to support the STS endpoints. 

The second part of this support eventually will go away.

> Please also note that this is an emerging space and the endpoints may change eventually before productization. More details about CIAM can be found here: [CIAM previews](https://github.com/microsoft/entra-previews)

**_Completed :_**
- Code changes: Added `AuthorityType: Ciam` and the path tranformation logic stated above
- Unit tests
- Docs updates
- Samples tested

**_Coming soon :_**
- e2e tests will be in a separate PR

### CODE REVIEW

File: lib/msal-common/src/authority/AuthorityFactory.ts
Comment (tnorling) [inline lib/msal-common/src/authority/AuthorityFactory.ts:40]: Can we do this in the authority constructor/instance methods instead?

File: lib/msal-common/src/authority/Authority.ts
Comment (tnorling) [inline lib/msal-common/src/authority/Authority.ts]: nit to make this a little less verbose:

```suggestion
     * If no path segments found and it is a CIAM authority (hostname ends with .ciamlogin.com), then transform it
```

File: lib/msal-common/src/authority/AuthorityFactory.ts
Comment (sameerag) [reply lib/msal-common/src/authority/AuthorityFactory.ts:40]: Guess we can. I chose this as most of the `authority` instance methods are getters and setters and having this code here made it more readable IMHO than in the other class. Let me wait for a few more reviews and make a call.

File: lib/msal-common/src/authority/Authority.ts
Comment (sameerag) [reply lib/msal-common/src/authority/Authority.ts]: Sorry, was meaning to fix this, pasted verbatim for a discussion as a place holder :). Committed the suggestion. Doc changes and samples are in progress for end users.

File: lib/msal-common/src/authority/AuthorityFactory.ts
Comment (tnorling) [reply lib/msal-common/src/authority/AuthorityFactory.ts:40]: The argument for doing it in `Authority` is that we may not always go through the `AuthorityFactory` in which case this would be missed. I'll leave it up to you but it feels safer to do important things like this in the `Authority` class itself.

**Comment by tnorling** [comment]
Changefile?

File: lib/msal-common/src/authority/AuthorityFactory.ts
Comment (sameerag) [reply lib/msal-common/src/authority/AuthorityFactory.ts:40]: I expect this to go away. The transition is temporary until STS supports the standard format without needing to append anything. So leaving as is for now.

File: lib/msal-common/src/authority/Authority.ts
Comment (derisen) [inline lib/msal-common/src/authority/Authority.ts]:
```suggestion
        // CIAM auth url pattern is being standardized as: <tenant>.ciamlogin.com
        if (this.canonicalAuthorityUrlComponents.HostNameAndPort.endsWith(Constants.CIAM_AUTH_URL)) {
```

File: lib/msal-common/src/authority/Authority.ts
Comment (derisen) [inline lib/msal-common/src/authority/Authority.ts]:
```suggestion
        if (this.authorityType === AuthorityType.Ciam) {
```

File: lib/msal-common/src/authority/Authority.ts
Comment (derisen) [inline lib/msal-common/src/authority/Authority.ts]:
```suggestion
        if (authorityUrlComponents.PathSegments.length === 0 && (authorityUrlComponents.HostNameAndPort.endsWith(Constants.CIAM_AUTH_URL))){
```

File: lib/msal-common/src/utils/Constants.ts
Comment (derisen) [inline lib/msal-common/src/utils/Constants.ts]: > CIAM_ENDPOINT: ".onmicrosoft.com",

super nit: maybe something like `AAD_TENANT_DOMAIN_SUFFIX` would be more accurate

File: lib/msal-common/test/authority/AuthorityFactory.spec.ts
Comment (derisen) [inline lib/msal-common/test/authority/AuthorityFactory.spec.ts]: is the test here aptly named? (seems like trailing slash is added)

File: lib/msal-common/test/authority/AuthorityFactory.spec.ts
Comment (derisen) [inline lib/msal-common/test/authority/AuthorityFactory.spec.ts]: same comment here: seems like it does not transform authority here

**Comment by derisen** [comment]
[.NET impl.](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4054) does not seem to allow setting CIAM authority at request level, though this doesn't seem to create an issue in here. Will test a bit more and confirm

**Comment by derisen** [comment]
> [.NET impl.](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4054) does not seem to allow setting CIAM authority at request level, though this doesn't seem to create an issue in here. Will test a bit more and confirm

Had another go at this. There isn't an issue from the service side, but we end up with the same situation as in B2C user-flows: multiple set of entries in the cache, making it complicated to maintain the right account for UI and token requests (without a good use case it seems to me, unlike B2C). 

Should we align the SDKs here? cc @trwalke for more context

File: lib/msal-common/src/utils/Constants.ts
Comment (sameerag) [reply lib/msal-common/src/utils/Constants.ts]: Sure. Can change this.

File: lib/msal-common/test/authority/AuthorityFactory.spec.ts
Comment (sameerag) [reply lib/msal-common/test/authority/AuthorityFactory.spec.ts]: No, this is a miss, I was moving around tests and did not rename. Will change.

**Comment by sameerag** [comment]
> > [.NET impl.](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4054) does not seem to allow setting CIAM authority at request level, though this doesn't seem to create an issue in here. Will test a bit more and confirm
> 
> Had another go at this. There isn't an issue from the service side, but we end up with the same situation as in B2C user-flows: multiple set of entries in the cache, making it complicated to maintain the right account for UI and token requests (without a good use case it seems to me, unlike B2C).
> 
> Should we align the SDKs here? cc @trwalke for more context

Sure. The reason why I have transformed it before and not later is that it is easy to clear this code once this is no longer needed. Can you please elaborate on multiple entries here? You mean all authorities of the format `*.ciamlogin.com/*` with the same `tenant` has to considered the same authority?

**Comment by derisen** [comment]
> > > [.NET impl.](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4054) does not seem to allow setting CIAM authority at request level, though this doesn't seem to create an issue in here. Will test a bit more and confirm
> > 
> > 
> > Had another go at this. There isn't an issue from the service side, but we end up with the same situation as in B2C user-flows: multiple set of entries in the cache, making it complicated to maintain the right account for UI and token requests (without a good use case it seems to me, unlike B2C).
> > Should we align the SDKs here? cc @trwalke for more context
> 
> Sure. The reason why I have transformed it before and not later is that it is easy to clear this code once this is no longer needed. Can you please elaborate on multiple entries here? You mean all authorities of the format `*.ciamlogin.com/*` with the same `tenant` has to considered the same authority?

No I mean, if I sign-in with `authority1`, and then later acquire a token for `authority2`, I would end up with a set of entries (account and tokens) in the cache for each. This is normal behavior of our SDK, regardless of AAD, B2C or CIAM. But I'm not sure why [.NET doesn't allow overriding authority in the case of CIAM](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4054/files#r1161000570) -hopefully @trwalke can step in to elaborate.

**Comment by sameerag** [comment]
> Changefile?

Added

**Comment by sameerag** [comment]
> > > > [.NET impl.](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4054) does not seem to allow setting CIAM authority at request level, though this doesn't seem to create an issue in here. Will test a bit more and confirm
> > > 
> > > 
> > > Had another go at this. There isn't an issue from the service side, but we end up with the same situation as in B2C user-flows: multiple set of entries in the cache, making it complicated to maintain the right account for UI and token requests (without a good use case it seems to me, unlike B2C).
> > > Should we align the SDKs here? cc @trwalke for more context
> > 
> > 
> > Sure. The reason why I have transformed it before and not later is that it is easy to clear this code once this is no longer needed. Can you please elaborate on multiple entries here? You mean all authorities of the format `*.ciamlogin.com/*` with the same `tenant` has to considered the same authority?
> 
> No I mean, if I sign-in with `authority1`, and then later acquire a token for `authority2`, I would end up with a set of entries (account and tokens) in the cache for each. This is normal behavior of our SDK, regardless of AAD, B2C or CIAM. But I'm not sure why [.NET doesn't allow overriding authority in the case of CIAM](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/4054/files#r1161000570) -hopefully @trwalke can step in to elaborate.

Discussed with @jmprieur and I believe we have it right. 

@jmprieur can you please review this too?

File: lib/msal-common/src/utils/Constants.ts
Comment (sameerag) [reply lib/msal-common/src/utils/Constants.ts]: Done

File: lib/msal-common/test/authority/AuthorityFactory.spec.ts
Comment (sameerag) [reply lib/msal-common/test/authority/AuthorityFactory.spec.ts]: Done

File: lib/msal-common/test/authority/AuthorityFactory.spec.ts
Comment (sameerag) [reply lib/msal-common/test/authority/AuthorityFactory.spec.ts]: Done

**Review by tnorling — APPROVED** [review]

File: samples/msal-browser-samples/VanillaJSTestApp2.0/app/ciam/index.html
Comment (derisen) [inline samples/msal-browser-samples/VanillaJSTestApp2.0/app/ciam/index.html]:
```suggestion
<h5 class="card-title" id="WelcomeMessage">Please sign-in</h5>
```

Comment (derisen) [reply samples/msal-browser-samples/VanillaJSTestApp2.0/app/ciam/index.html]: forgot that this will be replaced with username once they sign-in

**Review by derisen — APPROVED** [review]
LGTM!

**Pull request merged**
